Bio/Entrez/__init__.py

```diff
@@ -563,28 +558,47 @@ def parse(handle, validate=True, escape=False):
     return records
 
 
-def _open(cgi, params=None, post=None, ecitmatch=False):
-    """Build the URL and open a handle to it (PRIVATE).
-
-    Open a handle to Entrez.  cgi is the URL for the cgi script to access.
-    params is a dictionary with the options to pass to it.  Does some
-    simple error checking, and will raise an IOError if it encounters one.
-
-    The argument post should be a boolean to explicitly control if an HTTP
-    POST should be used rather an HTTP GET based on the query length.
-    By default (post=None), POST is used if the URL encoded parameters would
-    be over 1000 characters long.
-
-    This function also enforces the "up to three queries per second rule"
-    to avoid abusing the NCBI servers.
+def _open(req_or_cgi, params=None, post=None, ecitmatch=False):
+    """Make an HTTP request to Entrez, handling errors and enforcing rate limiting (PRIVATE).
+
+    Does some simple error checking and will try again after certain types of errors, up to
+    ``max_retries`` times. This function also enforces the "up to three queries per second
+    rule" to avoid abusing the NCBI servers (this limit is increased to 10 if using an API key).
+
+    This function previously also built the request from the CGI URL and parameter dictionary, that
+    functionality has since been moved to the ``_build_request()`` function. The old behavior has
+    been retained for backward compatibility but is now deprecated.
+
+    :param req_or_cgi: A Request object returned by ``_build_request``, or a URL to be passed
+        as the first argument to ``_build_request`` (deprecated).
+    :type req_or_cgi: urllib.request.Request or str
+    :param dict params: (Deprecated) a dictionary of options to be passed to ``_build_request`` if
+        the first argument is a string, ignored otherwise.
+    :param bool post: (Deprecated) whether to use the HTTP POST method instead of HTTP get. Passed
+        to ``_build_request`` if the first argument is a string, ignored otherwise.
+    :param bool ecitmatch: (Deprecated) passed to ``_build_request`` if the first argument is a
+        string, ignored otherwise.
+    :returns: Handle to HTTP response as returned by ``urllib.request.urlopen``. Will be wrapped in
+        an ``io.TextIOWrapper`` if its content type is plain text.
+    :rtype: http.client.HTTPResponse or io.TextIOWrapper
+    :raises urllib.error.URLError: Errors raises by ``urlopen`` past the maximum number of retries.
     """
+    if isinstance(req_or_cgi, Request):
+        request = req_or_cgi
+    else:
+        warnings.warn(
+            "Passing anything other than a Request object is deprecated, use _build_request() first.",
```

Comment:
Do we need this for a private function?

Any reason not to just change the private API?

Reply:
I don't think there is at all, but I wanted to check.

Reply:
Your proposal is more cautious, which is good. I'll leave this to @mdehoon to decide on.

Reply:
I would either raise a `RuntimeError`, or don't check at all.
Warnings are inconvenient because they are issued only once, and they don't tell you where the error occurred.
So you may be running a long script and suddenly
`Passing anything other than a Request object is deprecated, use _build_request() first.`
pops up, and you won't be able to (a) see where the error happened as there is no traceback, (b) replicate it.

Reply:
Removed this entirely.